---
title: "04 - GenomicDataCommons Normal Vs. Cancer tissue tRNA gene methylation - tRNAge sets & meth by tRNA"
author: "Richard J. Acton"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    fig_caption: yes
    number_sections: yes
    toc: yes
    toc_float: yes
  html_notebook:
    fig_caption: yes
    number_sections: yes
    toc: yes
    toc_float: yes
bibliography: "`r normalizePath('../library.bib')`"
---

# Set-up

## libs

```{r}
suppressPackageStartupMessages({
	library(tidyverse)
	library(lubridate)
	library(broom)
	library(psych)
	library(heatmaply)
	library(patchwork)
})
```

## Data Read-In

### tRNA array probes

```{r}
tRNAprobes <- 
read_delim(	
	delim = "\t",
	file = "../tRNA-GtRNAdb/450k_coresponding_hg19tRNAs.bed",
	col_types = "ciicciiciciicccc",
	col_names = c("pChr","pStart","pEnd","probeID",
		as.character(
			read_delim(
				delim = "\t",
				file = "../tRNA-GtRNAdb/std_tRNA_header.txt",
				col_names = FALSE,
				col_types = paste0(rep("c",12),collapse = "")
			)
		)
	)
)

```

### tRNAge gene set

```{r}
tRNAge <- read_delim(
	delim = "\t",
	 col_names = "tRNAname",
	 col_types = "c",
	 file = "../tRNA-GtRNAdb/GenWideSigWintRNAs.txt"
) %>%
	pull()
```

### Methylation Data

```{r}
dataTest <- readRDS(
	file = "data/tRNAprobesNormCancerArrayPairs.Rds"
)
```

```{r}
tRNAmethCancerNormal <- dataTest %>% 
	unnest() %>% 
	as_tibble() %>%
	filter(sample_type %in% c("Solid Tissue Normal","Primary Tumor")) %>%
	mutate(tRNAge = tRNAname %in% tRNAge) %>%
	filter(!is.na(Beta_value))

```

# tRNA subsets

```{r}
tRNAmethCancerNormal %>%
	mutate(tRNAge=as.character(tRNAge)) %>%
	mutate(tRNAge=if_else(tRNAge==TRUE,"tRNAge (MeDIP-seq 38)","Other tRNAs")) %>%
	ggplot(aes(sample_type,Beta_value)) + 
		geom_jitter(size=0.01) +
		geom_violin(aes(fill=sample_type),alpha=0.6,draw_quantiles = c(0.25,0.5,0.75)) +
		facet_wrap(~tRNAge)
```

```{r}
tRNAsCovered <- tRNAmethCancerNormal %>% distinct(tRNAname) %>% pull()
tRNAsCovered
tRNAsCovered %>% length()
```

```{r}
tRNAgeCovered <- tRNAge[tRNAge %in% tRNAsCovered]
tRNAgeCovered
tRNAgeCovered %>% length()
```

## tRNAge/other genes methylation with age in cancer and normal tissues

```{r,fig.width=12,fig.height=6.75}
tRNAgeCancerNormal <- 
tRNAmethCancerNormal %>% 
	mutate(tRNAge = as.character(tRNAge)) %>%
	mutate(
		tRNAge = if_else(
			tRNAge==TRUE,
			"tRNAge (MeDIP-seq 38)",
			"Other tRNAs"
		)
	) %>%
	ggplot(aes(age,Beta_value)) + 
		geom_point(size = 0.01) + 
		geom_smooth(method = "lm") + 
		geom_text(
			y = 0.94, x = 18, colour = "red",
			aes(label = paste0("r = ", round(r,4))),
			data = tRNAmethCancerNormal %>%
				mutate(tRNAge = as.character(tRNAge)) %>%
				mutate(tRNAge = if_else(
					tRNAge == TRUE,
					"tRNAge (MeDIP-seq 38)",
					"Other tRNAs")
				) %>%
				group_by(tRNAge,sample_type) %>%
				summarise(r = cor(age,Beta_value))
		) +
		facet_grid(sample_type ~ tRNAge) + 
		labs(title = "tRNA gene methylation with age accross all tissues")

tRNAgeCancerNormal
```

```{r}
#dir.create("../graphics")
# ggsave(tRNAgeCancerNormal,filename = "../graphics/tRNAgeCancerNormal.png",width = 12,height = 6.75)
```

## tRNAge/other genes methylation with age by tissue of origin

```{r,fig.width=12,fig.height=6.75}
ageRange <- range(tRNAmethCancerNormal$age)

rsByTissue <- tRNAmethCancerNormal %>%
	mutate(tRNAge=as.character(tRNAge)) %>%
	mutate(tRNAge=if_else(tRNAge==TRUE,"tRNAge (MeDIP-seq 38)","Other tRNAs")) %>%
	group_by(tRNAge,sample_type,primary_site) %>%
	summarise(r=cor(age,Beta_value))

tRNAgeCancerNormalbyTissue <- 
tRNAmethCancerNormal %>% 
	mutate(tRNAge=as.character(tRNAge)) %>%
	mutate(tRNAge=if_else(tRNAge==TRUE,"tRNAge (MeDIP-seq 38)","Other tRNAs")) %>%
	group_by(primary_site) %>%
	do(plot=
		ggplot(.,aes(age,Beta_value)) + 
			geom_point(size=0.01) + 
			geom_smooth(method = "lm") + 
			xlim(ageRange) +
			ylim(0,1) +
			geom_text(y=0.94,x=18,colour="red",  #####!!!!!!!
					  aes(label=paste0("r = ",round(r,4))),
				data = rsByTissue[rsByTissue$primary_site==.$primary_site[1],]
			) +
			facet_grid(sample_type~tRNAge) + 
			labs(title = paste0("tRNA gene methylation in ",.$primary_site))
	)

tRNAgeCancerNormalbyTissue$plot
```

---

# tRNA-iMet-CAT-1-4

much higher mean methylation in these samples than in fetal samples (~0.125), quite a lot of variance here especially in the tumour

```{r}
tRNAmethCancerNormal %>%
	dplyr::filter(tRNAname == "tRNA-iMet-CAT-1-4") %>%
	ggplot(aes(age,Beta_value)) +
		geom_point(aes(shape=sample_type,colour=sample_type)) + 
		scale_shape_manual(values = c(3,4)) +
		geom_smooth(aes(colour=sample_type),method = "lm") + 
		scale_colour_manual(values = c("red","black")) +
		facet_wrap(~tRNAname)
```

---

extracting codon/aa/ncular mitochondrial data from tRNA names

```{r}
tRNAmethCancerNormal <- 
tRNAmethCancerNormal %>%
	tidyr::extract(tRNAname,
			c("nmt","aa","codon"),
			"(\\w*)-?tRNA-(i?\\w{3})(?:\\w+)?-(\\w+)-",
			remove = FALSE
	)
```

```{r,fig.width=8,fig.height=4.5}
# custom labeler: (character vector in and out)
tRNA_labeller <- function(value) {sapply(value,
	function(n){
	tRNAmethCancerNormal %>%
		dplyr::select(tRNAname,tChr,strand) %>%
		distinct() %>%
		dplyr::filter(tRNAname==as.character(n)) %>%
		unlist() %>%
		(function(d) paste0(d["tRNAname"]," (",d["strand"],") ",tools::toTitleCase(d["tChr"]))) %>%
		return()
	}
)}

plots <- tRNAmethCancerNormal %>%
	group_by(aa) %>%
	do(plot=
		ggplot(.,aes(age,Beta_value)) +
			geom_point(aes(shape=sample_type,colour=sample_type)) + 
			scale_shape_manual(values = c(3,4)) +
			geom_smooth(aes(colour=sample_type),method = "lm") + 
			scale_colour_manual(values = c("red","black")) +
			facet_wrap(~tRNAname, labeller = labeller(tRNAname = tRNA_labeller)) +
			labs(title = paste0("tRNA genes of ",.$aa," isotype"),
				 x="Age /yrs"
			)
	)
plots$plot
```

```{r}
# #dir.create("graphics/tRNAMethAgeCancerVsNormalBytRNA")
# nil <- plots %>% do(out=ggsave(plot = .$plot,
# 					filename = paste0("graphics/tRNAMethAgeCancerVsNormalBytRNA/tRNAMethAgeCancerVsNormal_" , .$aa , ".png"),
# 					width = 8,
# 					height = 4.5
# 				))
# rm(nil)
```

```{r, eval = FALSE}
plots <- tRNAmethCancerNormal %>%
	group_by(aa,primary_site) %>%
	do(plot=
		ggplot(.,aes(age,Beta_value)) +
			geom_point(aes(shape=sample_type,colour=sample_type)) + 
			scale_shape_manual(values = c(3,4)) +
			geom_smooth(aes(colour=sample_type),method = "lm") + 
			scale_colour_manual(values = c("red","black")) +
			facet_wrap(~tRNAname, labeller = labeller(tRNAname = tRNA_labeller)) +
			labs(title = paste0("tRNA genes of ",.$aa," isotype for ",.$primary_site),
				 x="Age /yrs"
			)
	)
#plots$plot
```

```{r}
# # dir.create("graphics/tRNAMethAgeCancerVsNormalByTissue")
# nil <- plots %>% do(out=ggsave(plot = .$plot,
# 					filename = paste0("graphics/tRNAMethAgeCancerVsNormalByTissue/tRNAMethAgeCancerVsNormal_" , .$aa ,"_",.$primary_site, ".png"),
# 					width = 8,
# 					height = 4.5
# 				))
# rm(nil)
```

# Session Info

```{r}
sessionInfo()
```

# References

