---
title: "02 mouse-tRNA Age modeling"
author: "Richard J. Acton"
date: "`r Sys.Date()`"
output:
  html_document:
    fig_caption: yes
    number_sections: yes
    toc: yes
    toc_float: yes
    df_print: paged
    keep_md: yes
  html_notebook:
    fig_caption: yes
    number_sections: yes
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
    fig_caption: yes
    df_print: kable
bibliography: "`r normalizePath(params$bib)`"
csl: "`r normalizePath(params$csl)`"
link-citations: yes
linkcolor: blue
#runtime: shiny
editor_options: 
  chunk_output_type: inline
params:
  bib: "../library.bib"
  csl: "../genomebiology.csl"
---

# Introduction

Using mouse RRBS sequencing data from [@Petkovich2017] [GSE80672](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE80672) and mouse tRNA data from GtRNAdb [@Chan2009] to examine changes in tRNA gene DNAm with Age.

# Set-up

## Libs

```{r libs}
suppressPackageStartupMessages({
    library(GenomicRanges)
    library(BSgenome.Mmusculus.UCSC.mm10)
    library(TxDb.Mmusculus.UCSC.mm10.knownGene)
    library(tidyverse)
    library(plyranges)
    library(patchwork)
})
```

```{r}
if(!dir.exists("graphics")) {
    dir.create("graphics", showWarnings = FALSE, recursive = TRUE)
}
```


# Data Read-in

## tRNA

```{r}
mousetRNAs <- rtracklayer::import.bed("data/mm10-tRNAs.bed", genome = "mm10")
```

## Meth

```{r}
methData <- readRDS("data/tRNAmethdata.Rds") %>% 
    dplyr::select(-status, -submission_date, -type, -channel_count, -last_update_date)
```

## Blacklist

```{r}
mm10bl <- read_tsv(
    "data/mm10-blacklist.v2.bed.gz",
    col_names = c("seqnames","start","end","reason"),
    col_types = "ciic"
) %>% as_granges()
mm10bl
```

# Preprocessing

## tRNA data

removing tRx regions

```{r}
mousetRNAsSet <- 
mousetRNAs %>%
    as_tibble() %>%
    extract(
        name, into = c("tRNA", "aa", "anticodon", "f", "s"),
        regex = "(\\w+)-(\\w+)-(\\w+)-(\\w+)-(\\w+)",
        remove = FALSE
    ) %>% 
    filter(tRNA == "tRNA") %>%
    dplyr::select(-blocks) %>%
    as_granges()
```

```{r}
ntRNAs <- length(mousetRNAsSet)
ntRNAs

sws <- 0.05 / length(mousetRNAsSet)
sws
```

## meth data

```{r}
methDataDF <- 
methData %>% 
    mutate(data = map(data,as_tibble)) %>%
    unnest(cols = c(data))

# check base is CG
methDataDF$seq <- 
    Views(Mmusculus, Reduce("c", methData$data)) %>% 
    as("DNAStringSet") %>% 
    as.data.frame() %>%
    pull()

methDataGR <- methDataDF %>% plyranges::as_granges()
```

```{r}
nCpGsAvail <- 
    methDataGR %>% as_tibble() %>% distinct(seqnames,start) %>% nrow()
nCpGsAvail
```

## Combining tRNA and Meth data

```{r}
tRNAAndMeth <- join_overlap_intersect(mousetRNAsSet, methDataGR)
```

## Blacklist regions

tRNA covered

```{r}
tRNAAndMeth %>% as_tibble() %>% distinct(name) %>% nrow()
```

```{r}
tRNAAndMeth <- filter_by_non_overlaps(tRNAAndMeth, mm10bl)
```

tRNA covered - non-blacklist

```{r}
ntRNAsCovered <- 
tRNAAndMeth %>% as_tibble() %>% distinct(name) %>% nrow()
ntRNAsCovered
```

```{r}
propCovered <- ntRNAsCovered / ntRNAs
propCovered
```

```{r}
#Views(Mmusculus, methDataGR)
```

## GR to DF

```{r}
tRNAAndMethDF <- 
    tRNAAndMeth %>%
    as_tibble() %>%
    dplyr::rename(age = `age..years.`)
```

# Data properties

## treatment groups

```{r}
tRNAAndMethDF %>%
    distinct(geo_accession,diet) %>%
    group_by(diet) %>%
    dplyr::summarise(n = dplyr::n())
```

```{r}
tRNAAndMethDF %>%
    distinct(name,start) %>%
    nrow()
```

```{r}
tRNAAndMethDF <- 
tRNAAndMethDF %>%
    dplyr::filter(diet == "standard")
```

## Age distribution

```{r}
tRNAAndMethDF %>%
    distinct(geo_accession, age) %>%
    pull(age) %>%
    summary()
```


```{r}
tRNAAndMethDF %>%
    distinct(geo_accession, age) %>%
    ggplot(aes(age)) + 
        geom_density()
```

## tRNAs covered

```{r}
tRNAAndMethDF %>% 
    group_by(name) %>%
    dplyr::summarise(n=dplyr::n(), ageRange = diff(range(age)))
```

## Coverage distribution

```{r}
tRNAAndMethDF %>%
    dplyr::filter(diet == "standard") %>%
    ggplot(aes(cov)) + 
        geom_density(
            aes(group = geo_accession)
        )
```

```{r}
tRNAAndMethDF %>%
    ggplot(aes(cov)) + 
        geom_density(
            #aes(group = geo_accession)
        )
```

```{r}
tRNAAndMethDF %>% 
    dplyr::filter(cov >= 50) %>%
    group_by(name) %>%
    #dplyr::filter(dplyr::n() > 10) %>%
    dplyr::summarise(n=dplyr::n(), ageRange = diff(range(age)))
```

```{r}
QCtRNAset <- 
tRNAAndMethDF %>% 
    dplyr::filter(cov >= 50) %>%
    group_by(name) %>%
    dplyr::filter(dplyr::n() > 10) %>%
    dplyr::summarise(n=dplyr::n(), ageRange = diff(range(age)))

nQCtRNAset <- QCtRNAset %>% nrow()
nQCtRNAset
```

```{r}
propQCtRNAs <- nQCtRNAset / ntRNAs
propQCtRNAs
```

## Methylation distribution

```{r}
tRNAAndMethDF %>%
    ggplot(aes(meth)) + 
        geom_density(
            #aes(group = geo_accession)
        )
```

```{r}
tRNAAndMethDF %>%
    dplyr::filter(cov >= 50) %>%
    ggplot(aes(meth)) + 
        geom_density(
            #aes(group = geo_accession)
        )
```

# Age modeling

## by tRNA

```{r}
quickAgeModels <- 
tRNAAndMethDF %>%
    dplyr::filter(cov >= 50) %>%
    group_by(name) %>%
    nest() %>%
    mutate(model = map(data, ~lm(age ~ meth, data = .x))) %>%
    mutate(tidy = map(model, broom::tidy)) %>%
    unnest(cols = c(tidy)) %>%
    mutate(
        term = case_when(
            term == "(Intercept)" ~ "intercept",
            term == "meth" ~ "slope",
        )
    ) %>%
    dplyr::select(-std.error, -statistic, -p.value) %>%
    spread(term,estimate) %>%
    mutate(glance = map(model, broom::glance)) %>%
    unnest(cols = c(glance)) %>%
    arrange(p.value)

quickAgeModelssws <- 
quickAgeModels %>% 
    dplyr::select(-data, -model) %>%
    filter(p.value < sws)
quickAgeModelssws
```

```{r}
quickAgeModels %>%
    group_by(sign(slope)) %>%
    dplyr::summarise(n = dplyr::n())
```

```{r}
quickAgeModels %>%
    filter(p.value < 0.05) %>%
    group_by(sign(slope)) %>%
    dplyr::summarise(n = dplyr::n())
```

# DNAm age plots

## global age ~ DNAm

```{r}
tRNAAndMethDF %>%
    dplyr::filter(cov >= 50) %>%
        ggplot(aes(age,meth)) + 
            geom_point() +
            geom_smooth(method = "lm") 
```

## age ~ DNAm for sws tRNAs

```{r}
tRNAAndMethDF %>%
    dplyr::filter(cov >= 50) %>%
    dplyr::filter(name %in% quickAgeModelssws$name) %>%
    group_by(name) %>%
        ggplot(aes(age,meth)) + 
            geom_point() +
            geom_smooth(method = "lm") + 
            facet_wrap(~name)
```

## by CpG

```{r}
CpGsig <- 0.05 / nCpGsAvail
CpGsig
```

```{r}
quickAgeModelsCpG <- 
tRNAAndMethDF %>%
    dplyr::filter(cov >= 50) %>%
    group_by(name,start) %>%
    dplyr::filter(dplyr::n() > 10) %>%
    nest() %>%
    mutate(model = map(data, ~lm(age ~ meth, data = .x))) %>%
    mutate(tidy = map(model, broom::tidy)) %>%
    unnest(cols = c(tidy)) %>%
    mutate(
        term = case_when(
            term == "(Intercept)" ~ "intercept",
            term == "meth" ~ "slope",
        )
    ) %>%
    dplyr::select(-std.error, -statistic, -p.value) %>%
    spread(term,estimate) %>%
    mutate(glance = map(model, broom::glance)) %>%
    unnest(cols = c(glance)) %>%
    arrange(p.value)

quickAgeModelsCpGsws <- 
quickAgeModelsCpG %>% 
    dplyr::select(-data, -model) %>%
    filter(p.value < CpGsig) ### !!! acctual not total number of sites
quickAgeModelsCpGsws
```

```{r}
quickAgeModelsCpG %>%
    group_by(sign(slope)) %>%
    dplyr::summarise(n = dplyr::n())
```

```{r}
quickAgeModelsCpG %>%
    filter(p.value < 0.05) %>%
    group_by(sign(slope)) %>%
    dplyr::summarise(n = dplyr::n())
```

```{r}
quickAgeModelsCpG %>%
    dplyr::select(-data, -model) %>%
    filter(p.value < 0.05)
```

```{r}
byCpGplot <- function(df,modres) {
    
    addmod <- function(start) {
        sapply(start, function(srt){
            row <- modres %>% filter(start == srt)
            paste0(
                #sprintf("RÂ² = %.2f,\n p = %.2e\n", row$r.squared, row$p.value),
                "CpG: \n", format(as.numeric(srt), big.mark = ",")
            )
        })
    }
    
    df %>%
    ggplot(.,aes(age, meth)) + 
        geom_point(size = 0.8) +
        geom_smooth(aes(colour = `Significance`), method = "lm") + 
        facet_grid(
            paste0(
                gsub(name, pattern = "tRNA-", replacement = ""),
                " (", tools::toTitleCase(as.character(seqnames)), ")"
            ) ~ start,
            labeller = labeller(start = addmod),
            scales = "free"
        ) + 
        scale_colour_manual(values = c("darkblue","orange","red")) +
        scale_fill_discrete(drop=FALSE) + 
        theme_bw() + 
        theme(legend.position = "bottom") + 
        lims(y = c(0,85))+
        labs(
            x = "Age /months",
            y = "methylation /%"
        )
}
```

```{r fig.width=9, fig.height=3}
CpGPlots <- 
tRNAAndMethDF %>%
    dplyr::filter(cov >= 50) %>%
    dplyr::filter(name %in% quickAgeModelsCpGsws$name) %>%
    left_join(quickAgeModelsCpG,by = c("name","start")) %>%
    mutate(
        `Significance` = case_when(
            p.value < CpGsig ~ sprintf("p < %.3e",CpGsig),
            p.value < 0.05 ~ "p < 0.05",
            TRUE ~ "NS"
        )
    ) %>%
    group_by(name) %>%
    dplyr::filter(dplyr::n() > 10) %>%
    do(
        plots = byCpGplot(., quickAgeModelsCpG)
    ) #%>%
    #pull(plots)

#CpGPlots %>% pull(plots)
```

```{r, fig.width=9, fig.height= 7}
CpGPlotComb <- 
    {CpGPlots$plots[[1]] + theme(legend.position = "none")} /
    {CpGPlots$plots[[3]] + theme(legend.position = "none")} /
    CpGPlots$plots[[2]]
CpGPlotComb
```

```{r}
ggsave(
    CpGPlotComb,
    filename = "graphics/CpGPlotComb.png",
    width = 9,
    height = 7
)
```

# Session Info

```{r sessionInfo}
sessionInfo()
```

# References

<div id = "refs"></div>
